### PR TITLE
cli: WIP - REPL tab completion

### PR DESCRIPTION
**Pull request by shrirambalaji**
@MylesBorins I've made some minor progress for generating [tab completions](https://github.com/jstime/jstime/issues/10) for global builtins. In terms of the completions however, there are still some questions I have w.r.t Rustyline's API (https://github.com/kkawakam/rustyline/issues/441).

I've just started to adopt the `complete` function from node's repl into the [`generate_completions`](https://github.com/Shriram-Balaji/jstime/blob/c21d87d107a3a2e2f50a8a2b61605da1a3b869f1/cli/main.rs#L123) method - but I did notice a slightly different approach w.r.t how completions / annotations are built, in the newer node-prototype-repl which felt cleaner. Based on what I understand from [here](https://github.com/nodejs/repl/blob/db95cfb1a177b056b8bc257dfd62c7b0e50134df/src/index.js#L58), the acorn parser seems to parse the expression as an ESTree, and then we use that to correspondingly get the `propertyNames` / `functions` for a variable.

We could try to build a similar ESTree (not sure what lib would be suitable for that), (or) go with the regex-match-and-evaluate approach like the OG node repl. 

FWIW I'm still figuring my way out with Rustyline too, so the tab-completion itself isnt working, focusing primarily on getting the completions array filled up with the right annotations. 

cc @devsnek

### CODE REVIEW

File: cli/Cargo.toml [hunk 1]
```diff
@@ -18,6 +18,9 @@ jstime_core = { path = "../core", version = "^0.13.4-alpha.0"}
 rustyline = "6.3.0"
 structopt = "0.3.17"
 dirs = "3.0.1"
+rusty_v8 = "0.10.0"
```
Comment (devsnek) [inline cli/Cargo.toml:21]: why was this added?

File: cli/main.rs [hunk 2]
```diff
@@ -38,8 +49,12 @@ fn main() {
         "/snapshot_data.blob"
     )));
 
-    let mut jstime = jstime::JSTime::new(options);
+    JSTime::new(options)
+}
 
+fn main() {
+    let opt = Opt::from_args();
+    let mut jstime = init_runtime(&opt);
```
Comment (devsnek) [inline cli/main.rs:57]: seems like unrelated changes?

File: cli/main.rs [hunk 3]
```diff
@@ -49,49 +64,252 @@ fn main() {
...
+
+        rl.save_history(&self.history_file).unwrap();
     }
+}
+
+#[test]
 
-    if let Some(history_path) = history_path {
-        let _ = rl.save_history(&history_path);
+fn test_global_completions() {
```
Comment (devsnek) [inline cli/main.rs:223]: would be better to test the actual repl interface, like a file in the test dir which spawns jstime cli, sends some characters to stdin, and validates stdout.

**Review by devsnek — COMMENTED** [review]
it would be better to use the inspector. I believe rusty_v8 has the required inspector apis but i haven't really dug into it.

**Comment by Shriram-Balaji** [comment]
@devsnek The V8 inspector docs are apparently sparse according to rusty_v8: 
 https://docs.rs/rusty_v8/0.10.0/rusty_v8/inspector/index.html

Also, should jstime_core export something like a global context and scope? We could probably use this to directly make V8 calls from within the cli / repl ( fwiw took a peak at deno core, and they have some global_context it looks like in their JSRuntime )

**Comment by devsnek** [comment]
> should jstime_core export something like a global context and scope?

depends on what we need to run the inspector, i guess.

**Comment by MylesBorins** [comment]
We could also look at making an inspector api for jstime_code that hides
those internals

**Comment by Shriram-Balaji** [comment]
@MylesBorins i'll take a stab at implementing a facade of sorts for jstime and expose that as part of the core module. Inorder to avoid making it difficult to review, I'll probably do it in a separate PR instead.

**Comment by MylesBorins** [comment]
Quick ping, is this something we still want to pursue

**Comment by shrirambalaji** [comment]
I didn't had anytime to look into it now, post joining a new position. however I'll try to pick up something else which might need lesser amount of time.
